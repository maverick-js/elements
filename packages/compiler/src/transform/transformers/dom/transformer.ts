import {
  $,
  createImports,
  isJsxRootNode,
  resetUniqueIdCount,
  splitImportsAndBody,
  transformTsNode,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { Scope } from '../../../parse/ast';
import { parse } from '../../../parse/parse';
import { setupCustomElements } from '../shared/element';
import { Variables } from '../shared/variables';
import type { Transform, TransformData } from '../transformer';
import type { DomRuntime } from './runtime';
import { DomTransformState } from './state';
import { transform } from './transform';

export interface DomTransformOptions {
  /**
   * Whether to mark components with a static `element` property declaration to be
   * built as a custom element.
   */
  customElements?: boolean;
  /**
   * When set to `true`, the DOM runtime will connect the server-rendered HTML with the JS on
   * the client, making it interactive.
   *
   * The runtime does not require the component to be rendered on the server first but it's
   * recommended for the best performance.
   */
  hydratable?: boolean;
  /**
   * A single event listener is attached to the document to manage expensive events on multiple
   * child elements. Instead of adding individual listeners to each child, the event "bubbles" up
   * from the target element to the document, where the listener can capture and handle it. This
   * improves performance, especially in dynamic interfaces, by reducing the number of event
   * listeners.
   */
  delegateEvents?: boolean;
}

export function createDomTransform(options?: DomTransformOptions): Transform {
  return (data) => domTransform(data, options);
}

export function domTransform(
  { sourceFile }: TransformData,
  { customElements, hydratable, delegateEvents }: DomTransformOptions = {},
) {
  let state = new DomTransformState(null, { hydratable }),
    seenImports = false;

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast, new Scope())) : $.null;

      resetUniqueIdCount();

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  if (delegateEvents && state.delegatedEvents.size > 0) {
    body.push(createDelegateEventsStatement(state.delegatedEvents, state.runtime));
  }

  const statements: ts.Statement[] = [];

  const templates = createTemplateVariables(state);

  if (state.runtime.identifiers.length > 0) {
    statements.push(state.runtime.toImportDeclaration());
  }

  if (templates) statements.push(templates.toStatement());

  if (state.vars.module.length > 0) {
    statements.push(state.vars.module.toStatement());
  }

  const updatedSourceFile = $.updateSourceFile(transformedSourceFile, [
    ...imports,
    ...statements,
    ...state.renders,
    ...body,
  ]);

  return customElements
    ? setupCustomElements(updatedSourceFile, registerCustomElement, [createElementImport()])
    : updatedSourceFile;
}

const registerCustomElementExpression = $.call($.id('$$_define_custom_element'), [$.id('this')]);

function registerCustomElement() {
  return registerCustomElementExpression;
}

function createElementImport() {
  return createImports([$.id('$$_define_custom_element')], '@maverick-js/element');
}

function createTemplateVariables(state: DomTransformState) {
  const templates: [template: ts.Identifier, html: string, importNodes: boolean][] = [],
    vars = new Variables();

  state.walk(({ template, html, importNodes }) => {
    if (template && html) templates.push([template, html, importNodes]);
  });

  for (const [template, html, importNodes] of templates) {
    vars.create(template, state.runtime.createTemplate($.string(html), importNodes));
  }

  // Dedupe templates.
  for (let i = 0; i < templates.length; i++) {
    const templateA = templates[i];
    for (let j = i + 1; j < templates.length; j++) {
      const templateB = templates[j];
      if (templateA[1] === templateB[1] && templateA[2] === templateB[2]) {
        vars.update(templateB[0], templateA[0]);
      }
    }
  }

  return templates.length > 0 ? vars : null;
}

export function createDelegateEventsStatement(events: Set<string>, runtime: DomRuntime) {
  return $.createExpressionStatement(
    runtime.delegateEvents($.array(Array.from(events).map((type) => $.string(type)))),
  );
}
