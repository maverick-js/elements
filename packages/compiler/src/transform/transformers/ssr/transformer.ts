import { isArray } from '@maverick-js/std';
import {
  $,
  addStatementsToBlock,
  type BlockUpdate,
  createStatements,
  isJsxRootNode,
  replaceTsNodes,
  splitImportsAndBody,
  transformTsNode,
  updateBlock,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { parse } from '../../../parse/parse';
import { setupCustomElements } from '../shared/element';
import { returnLastExpression } from '../shared/factory';
import type { Transform, TransformData } from '../transformer';
import { SsrTransformState } from './state';
import { transform } from './transform';

export interface SsrTransformOptions {
  /**
   * Whether to mark components with a static `element` property declaration to be
   * rendered as a custom element.
   */
  customElements?: boolean;
}

export function createSsrTransform(options?: SsrTransformOptions): Transform {
  return (data) => ssrTransform(data, options);
}

export function ssrTransform(
  { sourceFile }: TransformData,
  { customElements }: SsrTransformOptions = {},
) {
  let state = new SsrTransformState(null),
    seenImports = false,
    blocks: BlockUpdate[][] = [];

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (ts.isBlock(node) || ts.isModuleBlock(node)) {
      const updates: BlockUpdate[] = [];

      blocks.push(updates);
      const block = ts.visitEachChild(node, visit, this);
      blocks.pop();

      return updateBlock(block, updates);
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast)) : $.null,
        currentBlock = blocks.at(-1);

      if (currentBlock) {
        if (isArray(result)) {
          const setup = result.slice(0, -1);

          currentBlock.push({
            type: 'add',
            statements: setup,
            before: node,
          });

          currentBlock.push({
            type: 'replace',
            old: node,
            new: result.at(-1)!,
          });
        } else {
          currentBlock.push({
            type: 'replace',
            old: node,
            new: result,
          });
        }

        // Return original block and transform above.
        return node;
      }

      if (isArray(result)) {
        returnLastExpression(result);
        return $.selfInvokedFn(result);
      }

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  const statements: ts.Statement[] = [];

  if (state.runtime.identifiers.length > 0) {
    statements.push(state.runtime.toImportDeclaration());
  }

  addTemplateVariables(state);
  if (state.vars.module.length > 0) {
    statements.push(state.vars.module.toStatement());
  }

  const updatedSourceFile = $.updateSourceFile(transformedSourceFile, [
    ...imports,
    ...statements,
    ...body,
  ]);

  return customElements
    ? setupCustomElements(updatedSourceFile, registerCustomElement)
    : updatedSourceFile;
}

function registerCustomElement() {
  return $.createTrue();
}

function addTemplateVariables(state: SsrTransformState) {
  state.walk(({ template, statics }) => {
    if (template && statics.length > 0) {
      state.vars.module.create(template, $.array(statics));
    }
  });
}
