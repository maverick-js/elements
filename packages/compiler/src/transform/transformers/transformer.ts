import type ts from 'typescript';

import type { AstNode } from '../../parse/ast';
import type { TransformContext } from '../transform';

export interface Transform {
  (data: TransformData): ts.SourceFile;
}

export interface TransformData {
  sourceFile: ts.SourceFile;
  nodes: AstNode[];
  ctx: TransformContext;
}

export interface StateTransform<State> {
  (node: AstNode, state: State): StateTransformResult;
}

export type StateTransformResult = ts.Expression | Array<ts.Expression | ts.Statement> | undefined;
