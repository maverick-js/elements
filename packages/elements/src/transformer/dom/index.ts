import { decode } from 'html-entities';
import MagicString from 'magic-string';
import * as t from 'typescript';
import {
  type AttributeNode,
  type ElementNode,
  isAttributeNode,
  isAttributesEnd,
  isDirectiveNode,
  isElementEnd,
  isElementNode,
  isEventNode,
  isExpressionNode,
  isRefNode,
  isSpreadNode,
  isStructuralNode,
  isTextNode,
  getAttrValue,
} from '../ast';
import { INNER_CONTENT_PROP } from '../jsx/constants';
import { type ASTSerializer } from '../transform';
import {
  createFunctionCall,
  createScopedDeclarations,
  createStringLiteral,
  trimWhitespace,
} from '../utils';

const spreadTrimRE = /(?:^\{\.{3})(.*)(?:\}$)/;

const ID = {
  template: '$$_templ',
  element: '$$_el',
  component: '$$_comp',
  markers: '$$_markers',
  expression: '$$_expr',
};

const RUNTIME = {
  createComponent: '$$_create_component',
  createTemplate: '$$_create_template',
  directive: '$$_directive',
  markers: '$$_markers',
  insert: '$$_insert',
  listen: '$$_listen',
  ref: '$$_ref',
  attr: '$$_attr',
  class: '$$_class',
  prop: '$$_prop',
  style: '$$_style',
  cssvar: '$$_cssvar',
  spread: '$$_spread',
  innerHTML: '$$_inner_html',
  innerText: '$$_inner_text',
  textContent: '$$_text_content',
  runHydrationEvents: '$$_run_hydration_events',
};

const MARKERS = {
  component: '<!C>',
  element: '<!E>',
  expression: '<!X>',
};

function createTemplate(template: string | string[], isSVG?: boolean) {
  return createFunctionCall(
    RUNTIME.createTemplate,
    [typeof template === 'string' ? template : template.join(''), isSVG && '1 /* SVG */'].filter(
      Boolean,
    ) as string[],
  );
}

export const dom: ASTSerializer = {
  serialize(ast, ctx) {
    let dynamicElCount = 0,
      currentId: string = '',
      markersId: string = '',
      delegate = false,
      templateId: string | undefined,
      expressions: string[] = [],
      template: string[] = [],
      element!: ElementNode,
      elements: ElementNode[] = [],
      globals = ctx.declarations,
      locals = createScopedDeclarations();

    // TODO: if component -> group props in an array (except ref/directive/events) + group all children?
    // map props to $$__create_props({ a:10, ..., get children() { return ...; } });
    // only children --> () => props.children

    const newMarkerId = () => {
      if (!markersId) {
        markersId = locals.create(ID.markers, createFunctionCall(RUNTIME.markers, [currentId]));
        ctx.runtimeImports.add(RUNTIME.markers);
      }
      return `${markersId}[${dynamicElCount++}]`;
    };

    const addAttrExpression = (node: AttributeNode, runtimeId: string) => {
      expressions.push(
        createFunctionCall(runtimeId, [
          currentId,
          createStringLiteral(node.name),
          getAttrValue(node),
        ]),
      );
      ctx.runtimeImports.add(runtimeId);
    };

    const addInnerContentExpression = (node: AttributeNode, runtimeId: string) => {
      expressions.push(createFunctionCall(runtimeId, [currentId, getAttrValue(node)]));
      ctx.runtimeImports.add(runtimeId);
    };

    for (const node of ast.tree) {
      if (isStructuralNode(node)) {
        if (isAttributesEnd(node)) {
          template.push('>');
        } else if (isElementEnd(node)) {
          element = elements.pop()!;
          if (!element?.isComponent) {
            template.push(element.isVoid ? `/>` : `</${element.tagName}>`);
            if (element.isSVG) template.push('</svg>');
          }
        }
      } else if (isElementNode(node)) {
        const dynamic = node.dynamic();

        if (dynamic) {
          if (!locals.has(ID.element)) {
            currentId = locals.create(
              ID.element,
              createFunctionCall(
                RUNTIME.createTemplate,
                [(templateId = globals.create(ID.template)), node.isSVG && '1 /* SVG */'].filter(
                  Boolean,
                ) as string[],
              ),
            );
            ctx.runtimeImports.add(RUNTIME.createTemplate);
          } else {
            currentId = locals.create(node.isComponent ? ID.component : ID.element, newMarkerId());
          }
        }

        if (node.isComponent) {
          template.push(MARKERS.component);
          expressions.push(
            createFunctionCall(RUNTIME.insert, [
              currentId,
              createFunctionCall(RUNTIME.createComponent, [node.tagName]),
            ]),
          );
          ctx.runtimeImports.add(RUNTIME.insert);
          ctx.runtimeImports.add(RUNTIME.createComponent);
        } else {
          if (node.isSVG) template.push('<svg>');
          if (dynamic && dynamicElCount > 0) template.push(MARKERS.element);
          template.push(`<${node.tagName}`);
        }

        element = node;
        elements.push(node);
      } else if (isAttributeNode(node)) {
        if (node.namespace) {
          if (node.namespace === '$attr') {
            addAttrExpression(node, RUNTIME.attr);
          } else if (node.namespace === '$class') {
            addAttrExpression(node, RUNTIME.class);
          } else if (node.namespace === '$style') {
            addAttrExpression(node, RUNTIME.style);
          } else if (node.namespace === '$cssvar') {
            addAttrExpression(node, RUNTIME.cssvar);
          } else {
            addAttrExpression(node, RUNTIME.prop);
          }
        } else if (INNER_CONTENT_PROP.has(node.name)) {
          if (node.name === '$innerHTML') {
            addInnerContentExpression(node, RUNTIME.innerHTML);
          } else if (node.name === '$innerText') {
            addInnerContentExpression(node, RUNTIME.innerText);
          } else if (node.name === '$textContent') {
            addInnerContentExpression(node, RUNTIME.textContent);
          }
        } else if (typeof node.value === 'string' || t.isStringLiteral(node.value)) {
          template.push(`${node.name}="${getAttrValue(node)}"`);
        } else if (node.name === 'value' || node.name === 'checked') {
          addAttrExpression(node, RUNTIME.prop);
        } else {
          addAttrExpression(node, RUNTIME.attr);
        }
      } else if (isRefNode(node)) {
        expressions.push(createFunctionCall(RUNTIME.ref, [currentId, node.value.getText()]));
        ctx.runtimeImports.add(RUNTIME.ref);
      } else if (isEventNode(node)) {
        const capture = node.namespace === '$oncapture';
        expressions.push(
          createFunctionCall(
            RUNTIME.listen,
            [
              currentId,
              createStringLiteral(node.type),
              node.value.getText(),
              (capture || node.delegate) && `${node.delegate ? 1 : 0} /* DELEGATE */`,
              capture && `1 /* CAPTURE */`,
            ].filter(Boolean) as string[],
          ),
        );

        if (node.delegate) {
          delegate = true;
          ctx.delegateEvents.add(node.type);
        }

        ctx.runtimeImports.add(RUNTIME.listen);
      } else if (isDirectiveNode(node)) {
        expressions.push(
          createFunctionCall(RUNTIME.directive, [currentId, node.name, node.value.getText()]),
        );
        ctx.runtimeImports.add(RUNTIME.directive);
      } else if (isTextNode(node)) {
        const text = decode(trimWhitespace(node.ref.getText()));
        if (text.length) template.push(text);
      } else if (isExpressionNode(node)) {
        if (typeof node.value === 'string') {
          template.push(node.value);
        } else {
          template.push(MARKERS.expression);

          const id = locals.create(ID.expression, newMarkerId());

          if (node.children) {
            let text = new MagicString(node.value.getText()),
              start = node.ref.getStart() + 1;

            for (const ast of node.children) {
              text.overwrite(
                ast.root.getStart() - start,
                ast.root.getEnd() - start,
                dom.serialize(ast, ctx),
              );
            }

            expressions.push(
              createFunctionCall(RUNTIME.insert, [
                id,
                node.observable ? `() => ${text}` : text.toString(),
              ]),
            );
          } else {
            expressions.push(
              createFunctionCall(RUNTIME.insert, [
                id,
                node.observable ? `() => ${node.value.getText()}` : node.value.getText(),
              ]),
            );
          }
        }
      } else if (isSpreadNode(node)) {
        expressions.push(
          createFunctionCall(
            RUNTIME.spread,
            [
              currentId,
              node.ref.getText().replace(spreadTrimRE, '$1'),
              (element.isSVG || element.hasChildren) && `${element.isSVG ? 1 : 0} /* SVG */`,
              element.hasChildren && '1 /* HAS_CHILDREN */',
            ].filter(Boolean) as string[],
          ),
        );
        ctx.runtimeImports.add(RUNTIME.spread);
      }
    }

    if (template.length) {
      const firstNode = ast.tree[0];
      const isSVG = isElementNode(firstNode) && firstNode.isSVG;
      const expression = createTemplate(template, isSVG);

      if (templateId) {
        globals.update(templateId, expression);
      } else {
        templateId = globals.create(ID.template, expression);
      }

      ctx.runtimeImports.add(RUNTIME.createTemplate);
    }

    if (locals.all.size) {
      if (delegate) ctx.runtimeImports.add(RUNTIME.runHydrationEvents);
      return [
        `(() => { `,
        `${locals.serialize()}`,
        ...expressions.join(';'),
        ';',
        delegate ? ` ${RUNTIME.runHydrationEvents}();` : '',
        ` return ${ID.element}; `,
        '})()',
      ].join('');
    } else {
      const firstNode = ast.tree[0];
      const isSVG = isElementNode(firstNode) && firstNode.isSVG;
      ctx.runtimeImports.add(RUNTIME.createTemplate);
      return createTemplate(templateId!, isSVG);
    }
  },
};
