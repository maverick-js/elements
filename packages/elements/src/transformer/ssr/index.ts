import type { ASTSerializer } from '../transform';

export const ssr: ASTSerializer = {
  serialize(ast, ctx) {
    return '';
  },
};
