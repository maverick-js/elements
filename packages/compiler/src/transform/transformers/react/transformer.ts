import { isArray } from '@maverick-js/std';
import {
  $,
  addStatementsToBlock,
  type BlockUpdate,
  isJsxRootNode,
  replaceTsNode,
  resetUniqueIdCount,
  splitImportsAndBody,
  transformTsNode,
  updateBlock,
} from '@maverick-js/ts';
import ts from 'typescript';

import {
  isMaverickCoreImportDeclaration,
  removeVirtualComponentImports,
} from '../../../parse/analysis';
import { parse } from '../../../parse/parse';
import { createDelegateEventsStatement } from '../dom/transformer';
import { returnLastExpression } from '../shared/factory';
import type { Transform, TransformData } from '../transformer';
import { ReactTransformState } from './state';
import { transform } from './transform';

export interface ReactTransformOptions {
  /**
   * A single event listener is attached to the document to manage expensive events on multiple
   * child elements. Instead of adding individual listeners to each child, the event "bubbles" up
   * from the target element to the document, where the listener can capture and handle it. This
   * improves performance, especially in dynamic interfaces, by reducing the number of event
   * listeners.
   */
  delegateEvents?: boolean;
}

export function createReactTransform(options?: ReactTransformOptions): Transform {
  return (data) => reactTransform(data, options);
}

export function reactTransform(
  { sourceFile }: TransformData,
  { delegateEvents }: ReactTransformOptions = {},
) {
  let state = new ReactTransformState(null),
    seenImports = false,
    blocks: BlockUpdate[][] = [];

  function visit(this: ts.TransformationContext, node: ts.Node) {
    if (!seenImports && isMaverickCoreImportDeclaration(node)) {
      seenImports = true;
      return removeVirtualComponentImports(node, (name) => state.runtime.add(name));
    } else if (ts.isBlock(node) || ts.isModuleBlock(node)) {
      const updates: BlockUpdate[] = [];

      blocks.push(updates);
      const block = ts.visitEachChild(node, visit, this);
      blocks.pop();

      return updateBlock(block, updates);
    } else if (isJsxRootNode(node)) {
      const ast = parse(node),
        result = ast ? transform(ast, state.child(ast)) : $.null,
        currentBlock = blocks.at(-1);

      resetUniqueIdCount();

      if (currentBlock) {
        if (isArray(result)) {
          const statements = result.slice(0, -1);

          currentBlock.push({
            type: 'add',
            statements,
            before: node,
          });

          currentBlock.push({
            type: 'replace',
            old: node,
            new: result.at(-1)!,
          });
        } else {
          currentBlock.push({
            type: 'replace',
            old: node,
            new: result,
          });
        }

        // Return original block and transform above.
        return node;
      }

      if (isArray(result)) {
        returnLastExpression(result);
        return $.selfInvokedFn(result);
      }

      return result;
    }

    return ts.visitEachChild(node, visit, this);
  }

  const transformedSourceFile = transformTsNode(sourceFile, (ctx) => visit.bind(ctx)),
    { imports, body } = splitImportsAndBody(transformedSourceFile);

  if (delegateEvents && state.delegatedEvents.size > 0) {
    body.push(createDelegateEventsStatement(state.delegatedEvents, state.domRuntime));
  }

  const statements: ts.Statement[] = [];

  const runtimes = [state.runtime, state.domRuntime, state.ssrRuntime];
  for (const runtime of runtimes) {
    if (runtime.identifiers.length > 0) {
      statements.push(runtime.toImportDeclaration());
    }
  }

  if (state.module.vars.length > 0) {
    statements.push(state.module.vars.toStatement());
  }

  return $.updateSourceFile(sourceFile, [
    ...imports,
    ...statements,
    ...state.module.block,
    ...body,
  ]);
}
