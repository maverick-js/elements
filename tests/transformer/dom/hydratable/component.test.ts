import { transform } from 'maverick.js/transformer';

const t = (code: string) => transform(code, { hydratable: true }).code;

it('should compile root component', () => {
  const result = t(`<Component />`);
  expect(result).toMatchSnapshot();
});

it('should compile child component', () => {
  const result = t(`<div><Component /></div>`);
  expect(result).toMatchSnapshot();
});

it('should compile root expression ', () => {
  const result = t(`
function Component() {
  return id > 10 ? <div>{id}</div> : 20;
}
  `);
  expect(result).toMatchSnapshot();
});

it('should compile nested components', () => {
  const result = t(`
<Component>
  Text
  <Foo>
    <div>{id()}</div>
  </Foo>
  <Bar />
</Component>
  `);
  expect(result).toMatchSnapshot();
});

it('should compile component with props', () => {
  const result = t(`<Component foo="a" bar={id()} />`);
  expect(result).toMatchSnapshot();
});

it('should compile component with spread', () => {
  const result = t(`<Component {...props} />`);
  expect(result).toMatchSnapshot();
});

it('should compile component with multiple spreads', () => {
  const result = t(`<Component {...props} {...propsTwo} {...propsThree} />`);
  expect(result).toMatchSnapshot();
});

it('should compile component with props and spread', () => {
  const result = t(`<Component foo="..." bar={id() + 10} {...props} />`);
  expect(result).toMatchSnapshot();
});

it('should compile component with text children', () => {
  const result = t('<Component>foo 10 bar 20 baz</Component>');
  expect(result).toMatchSnapshot();
});

it('should compile component with element children', () => {
  const result = t(`<Component><div>Foo{id()}</div></Component>`);
  expect(result).toMatchSnapshot();
});

it('should compile component with props and children', () => {
  const result = t(`<Component foo={id}><div></div></Component>`);
  expect(result).toMatchSnapshot();
});

it('should forward single call expression', () => {
  const result = t(`<Component>{() => <div>{id()}</div>}</Component>`);
  expect(result).toMatchSnapshot();
});

it('should forward multiple call expressions', () => {
  const result = t(`
<Component>
  {() => <div>{id()}</div>}
  {() => <div>{id()}</div>}
</Component>
`);
  expect(result).toMatchSnapshot();
});

it('should compile component child fragment', () => {
  const result = t(
    `
<Component>
  <>
  <div id="foo">{id()}</div>
  <div id="bar">{id()}</div>
  </>
</Component>
`,
  );
  expect(result).toMatchSnapshot();
});

it('should insert multiple child components', () => {
  const result = t(`
<div>
  <Component />
  <Component />
  <div>Foo</div>
  <Component />
</div>`);
  expect(result).toMatchSnapshot();
});

it('should insert fragmented child components', () => {
  const result = t(`
<div>
  <>
    <Component />
    <Component />
    <div>Foo</div>
    <Component />
  </>
</div>`);
  expect(result).toMatchSnapshot();
});
