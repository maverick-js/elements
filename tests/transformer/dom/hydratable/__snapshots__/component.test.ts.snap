// Vitest Snapshot v1

exports[`should compile child component 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_insert_at_marker, $$_create_component, $$_create_template } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker($$_comp, $$_create_component(Component));

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component containing expression 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_insert_at_marker, $$_create_template, $$_create_component } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_templ;
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker(
    $$_comp,
    $$_create_component(Component, {
      get children() {
        return () =>
          (() => {
            const $$_root = $$_clone($$_templ_2),
              $$_walker = $$_create_markers_walker($$_root),
              $$_expr = $$_walker.nextNode();

            $$_insert_at_marker($$_expr, id);

            return $$_root.firstChild;
          })();
      },
    }),
  );

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component with element children 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_create_template, $$_insert_at_marker, $$_create_component } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker(
    $$_comp,
    $$_create_component(Component, {
      get children() {
        return $$_clone($$_templ_2, 1 /* ELEMENT */);
      },
    }),
  );

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component with fragment children 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_create_template, $$_insert_at_marker, $$_create_component } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_create_template(\`<div id=\\"foo\\"></div><div id=\\"bar\\"></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker(
    $$_comp,
    $$_create_component(Component, {
      get children() {
        return $$_clone($$_templ_2);
      },
    }),
  );

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component with multiple spreads 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_insert_at_marker, $$_create_component, $$_merge_props, $$_create_template } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker($$_comp, $$_create_component(Component, $$_merge_props(props, propsTwo, propsThree)));

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component with props 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_insert_at_marker, $$_create_component, $$_create_template } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker($$_comp, $$_create_component(Component, { foo: \\"a\\", bar: id }));

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component with props and children 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_create_template, $$_insert_at_marker, $$_create_component } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`),
  $$_templ_2 = /* #__PURE__ */ $$_create_template(\`<div></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker(
    $$_comp,
    $$_create_component(Component, {
      foo: id,
      get children() {
        return $$_clone($$_templ_2, 1 /* ELEMENT */);
      },
    }),
  );

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component with props and spread 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_insert_at_marker, $$_create_component, $$_merge_props, $$_create_template } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker(
    $$_comp,
    $$_create_component(
      Component,
      $$_merge_props(props, {
        foo: \\"...\\",
        get bar() {
          return id() + 10;
        },
      }),
    ),
  );

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component with spread 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_insert_at_marker, $$_create_component, $$_merge_props, $$_create_template } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`);
(() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode();

  $$_insert_at_marker($$_comp, $$_create_component(Component, props));

  return $$_root.firstChild;
})()"
`;

exports[`should compile child component with text children 1`] = `
"import { $$_create_component } from \\"@maverick-js/elements/dom\\";
$$_create_component(Component, {
  get children() {
    return \\"foo 10 bar 20 baz\\";
  },
})"
`;

exports[`should compile root component 1`] = `
"import { $$_create_component } from \\"@maverick-js/elements/dom\\";
$$_create_component(Component)"
`;

exports[`should compile root expression  1`] = `
"import { $$_clone, $$_create_markers_walker, $$_insert_at_marker, $$_create_template } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$></div>\`);

function Component() {
  return id > 10 ? (() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_expr = $$_walker.nextNode();

  $$_insert_at_marker($$_expr, id);

  return $$_root.firstChild;
})() : 20;
}
  "
`;

exports[`should insert multiple child components 1`] = `
"import { $$_clone, $$_create_markers_walker, $$_insert_at_marker, $$_create_component, $$_create_template } from \\"@maverick-js/elements/dom\\";

const $$_templ = /* #__PURE__ */ $$_create_template(\`<div><!$><!$><div></div><!$></div>\`);

  (() => {
  const $$_root = $$_clone($$_templ),
    $$_walker = $$_create_markers_walker($$_root),
    $$_comp = $$_walker.nextNode(),
    $$_comp_2 = $$_walker.nextNode(),
    $$_comp_3 = $$_walker.nextNode();

  $$_insert_at_marker($$_comp, $$_create_component(Component));
  $$_insert_at_marker($$_comp_2, $$_create_component(Component));
  $$_insert_at_marker($$_comp_3, $$_create_component(Component));

  return $$_root.firstChild;
})()"
`;
